include/pybind11/detail/internals.h

```diff
@@ -280,21 +281,112 @@ inline internals **&get_internals_pp() {
     return internals_pp;
 }
 
+#if PY_VERSION_HEX >= 0x03030000
+inline bool raise_err(PyObject *exc_type, const char *msg) {
+    bool err_occurred = PyErr_Occurred() != nullptr;
+    if (err_occurred) {
+        raise_from(exc_type, msg);
+    } else {
+        PyErr_SetString(exc_type, msg);
+    }
+    return err_occurred;
+};
+
+// forward decl
+inline void translate_exception(std::exception_ptr);
+
+template <
+    class T,
+    enable_if_t<std::is_same<std::nested_exception, remove_cv_t<remove_reference_t<T>>>::value,
+                int> = 0>
+bool handle_nested_exception(const T &exc, const std::exception_ptr &p) {
+    std::exception_ptr nested = nullptr;
+    nested = exc.nested_ptr();
+    if (nested != nullptr && nested != p) {
+        translate_exception(nested);
+        return true;
+    }
+    return false;
+}
+
+template <
+    class T,
+    enable_if_t<!std::is_same<std::nested_exception, remove_cv_t<remove_reference_t<T>>>::value,
+                int> = 0>
+bool handle_nested_exception(const T &exc, const std::exception_ptr &p) {
+    if (auto *nep = dynamic_cast<const std::nested_exception *>(std::addressof(exc))) {
```

Comment:
Cooool!
I'd probably add
```
assert(nep != nullptr);
```
as a sanity check. (overhead only in debug mode, but even then it's negligible)

Reply:
`if` literally checks for this being nullptr, I don't think we need to add an assert too. If a reader doesn't understand the `if`, they might not understand the `assert` either. A comment could be added if you _really_ want it, but I'm not a fan of explaining C++ syntax in comments unless it's obscure or easy to miss - you can't miss a block. (If that's what you are referring to here).

Reply:
Oh, sorry, you're right. I was going too fast (basically I was overlooking the `if`).
I'd write this as
```
auto *nep = dynamic_cast<const std::nested_exception *>(std::addressof(exc));
if (nep != nullptr) {
```
because I believe that's less likely to get misread, but I'm totally fine with this code as is.